src/spi.rs

```diff
@@ -709,3 +804,132 @@ impl_spi!(SPI3: spi_host_device_t_SPI3_HOST);
 impl_spi_any_pins!(SPI2);
 #[cfg(not(esp32c3))]
 impl_spi_any_pins!(SPI3);
+
+use crate::gpio::{Level, Output};
+use crate::task::CriticalSection;
+use crate::delay::Ets;
+
+pub struct GuardedSpiDevice<'d, DRIVER: Borrow<SpiMasterDriver<'d>>> {
+    device: EspSpiDevice<'d, DRIVER>,
+    mutex: CriticalSection,
+}
+
+impl<'d, DRIVER> GuardedSpiDevice<'d, DRIVER>
+where
+    DRIVER: Borrow<SpiMasterDriver<'d>> + 'd,
+{
+    pub fn new(device: EspSpiDevice<'d, DRIVER>) -> Self {
+        Self {
+            device,
+            mutex: CriticalSection::new(),
+        }
+    }
+}
+pub struct EspSCSSpiDevice<'d, DEVICE, DRIVER: Borrow<SpiMasterDriver<'d>>> {
+    shared_device: DEVICE,
+    cs_pin: PinDriver<'d, AnyOutputPin, Output>,
+    pre_delay: Option<u32>,
+    post_delay: Option<u32>,
+    _p: PhantomData<&'d DRIVER>,
+}
+
+impl<'d, DEVICE, DRIVER> EspSCSSpiDevice<'d, DEVICE, DRIVER>
+where
+    DEVICE: Borrow<GuardedSpiDevice<'d, DRIVER>>,
+    DRIVER: Borrow<SpiMasterDriver<'d>>,
+{
+    pub fn new(
+        shared_device: DEVICE,
+        cs: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
+        cs_level: Level,
+    ) -> Result<Self, EspError> {
+        let cs_ref: PeripheralRef<AnyOutputPin> = cs.into_ref().map_into();
+        let mut cs_pin = PinDriver::output(cs_ref)?;
+        cs_pin.set_level(cs_level)?;
+        Ok(Self {
+            shared_device,
+            cs_pin,
+            _p: PhantomData,
+            pre_delay: None,
+            post_delay: None,
+        })
+    }
+    /// Add an aditional delay of x in uSeconds before transaction
+    /// between chip select and first clk out
+    pub fn cs_pre_delay(mut self, delay: u32) -> Self {
+        self.pre_delay = Some(delay);
+        self
+    }
+
+    /// Add an aditional delay of x in uSeconds after transaction
+    /// between last clk out and chip select
+    pub fn cs_post_delay(mut self, delay: u32) -> Self {
+        self.post_delay = Some(delay);
+        self
+    }
+
+
+    pub fn transaction<R, E>(
+        &mut self,
+        f: impl FnOnce(&mut SpiBusMasterDriver<'d>) -> Result<R, E>,
+    ) -> Result<R, E>
+    where
+        E: From<EspError>,
+    {
+        let shared_device: &GuardedSpiDevice<'d, DRIVER> = self.shared_device.borrow();
+        let _lock = shared_device.mutex.enter();
+        let device: &EspSpiDevice<'d, DRIVER> = shared_device.device.borrow();
+        self.cs_pin.toggle()?;
+        if let Some(delay) = self.pre_delay {
+            Ets::delay_us(delay);
+        }
```

Comment:
Considering the placement of these delays, couldn't the user just add this delay themselves in the `transaction` function? I don't see why it has do be done by `esp-idf-hal`.

Reply:
the point for this is, that if you are using an hardware library that expects an embedded_hal SpiDevice  you just can give it this Device's. The hardware driver manages the SpiDevice -> you have no way of hooking in the transaction yourself

Reply:
question become here if one writes an hardware driver for x, should he assume that the embedded_hal device that he receive is an devices with managed cs or not. If he specify that he would like to have an managed cs device and it needed an delay between cs and clk than you would need this.

Reply:
The e-hal traits do not assume or specify in any way how the SPI protocol is implemented. So as long as it is behaving as expected - managed CS or not - user should not care.  So I'm fine with these delays actually.

Reply:
> The hardware driver manages the SpiDevice -> you have no way of hooking in the transaction yourself

The nice thing about the `SpiDevice` trait is that it can be implemented by anyone. Which means anyone can simply make a wrapper to add this delay themselves and it's only a couple lines of code.

> question become here if one writes an hardware driver for x, should he assume that the embedded_hal device that he receive is an devices with managed cs or not

Yes. The Spi traits explicitly state that `SpiDevice` implementations take care of toggling the CS pin. If the driver wants to manage the CS pin then the `SpiBus` trait should be used instead.

Reply:
> This is not how I interpret the spec, sorry. SpiBus is here for bitbanging and all sorts of tricks on top of the SPI protocol that - besides - also require an exclusive access to the bus.

bitbanging is one use case. SD cards is another. Display+Touch screen combo drivers is yet another.
https://github.com/rust-embedded/embedded-hal/blob/master/embedded-hal/src/spi.rs#L96
It's all about whether exclusive access is required and it just so happens that software CS required exclusive access. (Hence the locking we have here and in esp-idf).
I think we're on the same page here but using different words.

> I think removing the CS pin delays, yet keeping the "pool" (what is now the `SpiSoftCsDeviceDriver` + `SpiSharedDeviceDriver` combo) is probably the worst. Either we provide "software" CS support, or we don't. But no in-between.

Oh okay, I kinda see what you're getting at.
First I want to say, does software CS actually require delays? Or is this just a convenience thing?
I see it as a convenience which is why I don't like it but from what you're saying it sounds like it required?

Though, now that the CS pin has been made optional, I think "software" CS feature should be removed then 😅. It should be easy for users to DIY it now.
A mutex (Critical section) is used here but some users might actually want `RefCell`. At the end of the day, increasing the device limit isn't about multithreading, it's about sharing, which Rust itself has multiple tools for.
So the code here is quite an opinionated feature.